controller_manager/src/controller_manager.cpp

```diff
@@ -2269,21 +2269,25 @@ controller_interface::return_type ControllerManager::update(
         run_controller_at_cm_rate ? period
                                   : rclcpp::Duration::from_seconds((1.0 / controller_update_rate));
 
+      const rclcpp::Time current_time = get_clock()->now();
       if (
-        *loaded_controller.next_update_cycle_time ==
+        *loaded_controller.last_update_cycle_time ==
         rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type()))
       {
         // it is zero after activation
+        *loaded_controller.last_update_cycle_time = current_time - controller_period;
         RCLCPP_DEBUG(
-          get_logger(), "Setting next_update_cycle_time to %fs for the controller : %s",
-          time.seconds(), loaded_controller.info.name.c_str());
-        *loaded_controller.next_update_cycle_time = time;
+          get_logger(), "Setting last_update_cycle_time to %fs for the controller : %s",
+          loaded_controller.last_update_cycle_time->seconds(), loaded_controller.info.name.c_str());
       }
+      const auto controller_actual_period =
+        (current_time - *loaded_controller.last_update_cycle_time);
 
-      bool controller_go =
+      const bool controller_go =
+        run_controller_at_cm_rate ||
         (time ==
          rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type())) ||
-        (time.seconds() >= loaded_controller.next_update_cycle_time->seconds());
+        (controller_actual_period.seconds() * controller_update_rate >= 0.99);
```

Comment:
is there a reason for 0.99 instead of 1.00? why exactly 1%?

Reply:
When tested using 1.0, when there is a jitter, even though you are close you are forced to wait for on more cycle and this are seen in peaks which are not nice enough. 

For instance, if you have a controller that runs at 50 Hz and the CM running at 100Hz, then when you have like 0.019 instead of triggering (as triggering should happen at 0.02), if you wait, then you trigger it at ~0.029 sec and this is not fair enough to keep up with the controller update rate

Reply:
i used a similar approach in PR: #1570 as we observed such issues.

Using  0.99 you will have the following (the plot is in nanoseconds)
![image](https://github.com/user-attachments/assets/47cda054-5085-4141-bf13-db8e74273591)

Using 1.0 you will have the following (the plot is in nanoseconds)
![image (12)](https://github.com/user-attachments/assets/d7a29a9c-7551-4b7f-bee5-a0eb254ab8ea)


In the above case, the CM is running at 1kHz and the components are running at 1000Hz, 100Hz, and 40Hz

Reply:
I was more referring to "greater **or equal**" vs "less **than**"

Reply:
So this is not a problem of non-integer divisors but of jitter etc?
I see, maybe you could add this explanation in the code, just to document that "magic number" :D

Reply:
Exactly, this is due to the jitter

I agree. I can add some explanation to the code.

Reply:
@christophfroehlich Done!